Added a gravity compensation package, modified sleep positions

Shiming-Liang opened the pull request:
- Added the gravity compensation demo package for xsarms
- Modified the sleep positions for aloha wx250s and vx300s so it doesn't fall as much

Shiming-Liang committed:
added demo of gravity compensation on xsarm

Shiming-Liang committed:
added config and launch files for gravity compensation demo on xsarm

Shiming-Liang committed:
updated doc for gravity compensation

Shiming-Liang committed:
properly formatted

Shiming-Liang committed:
move the config doc to interbotix_gravity_compensation

Shiming-Liang committed:
use namespace in the launch file

Shiming-Liang committed:
fixed typo in the grav comp yaml files, pointed to the closer example config

Shiming-Liang committed:
two arms can run at the same time, added parameters for friction modelling

Shiming-Liang committed:
corrected some entries of the URDF for  the aloha wx250s

Shiming-Liang committed:
updated the config file for more sophisticated friction compensation, updated the corresponding contents at the readme file, updated the URDF file for the aloha wx250s.

TO DO: update the link to the documentation of the  motor specs configuration.

Shiming-Liang committed:
fixed an incorrect doc link

Shiming-Liang committed:
changed the sleep positions so the arm doesn't drop as much

The arms involved: aloha wx250s, aloha vx300s

Shiming-Liang committed:
updated README for the gravity compensation package

- matched the node name with the one used in the code

Shiming-Liang committed:
Updated README.md for the interbotix_xsarm_gravity_compensation package

added link to the doc page with a example motor specs configuration

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/README.md:
```diff
@@ -0,0 +1,27 @@
+The arm will hold itself against gravity and can be moved freely when the gravity compensation is enabled.
+It will lock in its current position when the gravity compensation is disabled.
+
+**WARNING: the arm WILL torque off and drop for a short period of time while enabling/disabling. Please make sure it is in a resting position or manually held.**
+
+**WARNING: the joints not supporting current control WILL torque off. Please make sure to use arm with at least the first three joints supporting current control, e.g., RX, WX, VX series.**
```
For these warnings, we can use the GitHub-supported admonition style in markdown: https://github.com/orgs/community/discussions/16925

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/config/mode_configs_aloha_wx250s.yaml:
```diff
@@ -0,0 +1,21 @@
+# It should be placed in the 'config' directory of any ROS package that builds ontop of the interbotix_xs_sdk.
+# Example values are shown as well. All parameters are mandatory unless stated otherwise
+
+port: /dev/ttyDXL_awx250s                     # Optional; specifies the USB port that connects to the U2D2;
```
Keep the port as the standard `/dev/ttyDXL`

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/config/mode_configs_aloha_wx250s.yaml:
Mode configs can probably just be a single file. There's no difference between the wx250s and awx250s.

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/config/mode_configs_wx250s.yaml:
Remove this file per my previous comment

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/launch/interbotix_gravity_compensation.launch.py:
```diff
@@ -0,0 +1,132 @@
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            'robot_model',
+            choices=get_interbotix_xsarm_models(),
```
choices should just be `wx250s` and `aloha_wx250s`.

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/launch/interbotix_gravity_compensation.launch.py:
```diff
@@ -0,0 +1,132 @@
+                PathJoinSubstitution([
+                    FindPackageShare('interbotix_xsarm_gravity_compensation'),
+                    'config']),
+                "/motor_specs_",
```
This is a bit messy. Try to do something like this: https://github.com/Interbotix/interbotix_ros_manipulators/blob/552368e2ac0f8fdba69ee792ad71381238d78118/interbotix_ros_xsarms/examples/interbotix_xsarm_puppet/launch/xsarm_puppet.launch.py#L228-L231

Shiming-Liang replied:
Is this better?

```python
            default_value=[PathJoinSubstitution([
                FindPackageShare('interbotix_xsarm_gravity_compensation'),
                'config',
                "motor_specs_"]),
                LaunchConfiguration('robot_model'),
                '.yaml'
            ],
```

lukeschmitt-tr replied:
Would prefer the style below:

```python
default_value=[PathJoinSubstitution([
    FindPackageShare('interbotix_xsarm_gravity_compensation'),
    'config',
    'motor_specs_']), LaunchConfiguration('robot_model'), '.yaml'
],
```

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/package.xml:
```diff
@@ -0,0 +1,20 @@
+  <author email="trsupport@trossenrobotics.com">Shiming Liang</author>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+  <depend>interbotix_gravity_compensation</depend>
```
This also depends on interbotix_xsarm_control since it's being launched.

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/README.md:
```diff
@@ -0,0 +1,27 @@
+# interbotix_xsarm_gravity_compensation
+
+## Overview
+This package demos the interbotix_gravity_compensation package on an Interbotix arm.
```
Newline between headings and their body.

```suggestion
## Overview

This package demos the interbotix_gravity_compensation package on an Interbotix arm.
```

Shiming-Liang committed:
Minor fixes on gravity compensation

readme
- used the Github provided warning block
- added empty lines after headers

mode configs
- changed the port setting to the default /dev/ttyDXL
- merged the two mode configs file

launch
- limited choices to wx250s and aloha_wx250s
- improved readability for path concatenations

package.xml
- added dependency for interbotix_xsarm_control

Shiming-Liang committed:
Minor fixes on the gravity compensation package

- change file path concatenation style for better readability

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/config/mode_configs.yaml:
```diff
@@ -0,0 +1,21 @@
+# It should be placed in the 'config' directory of any ROS package that builds ontop of the interbotix_xs_sdk.
+# Example values are shown as well. All parameters are mandatory unless stated otherwise
+
+port: /dev/ttyDXL_awx250s                             # Optional; specifies the USB port that connects to the U2D2;
```
Keep this as `/dev/ttyDXL`

```suggestion
port: /dev/ttyDXL                             # Optional; specifies the USB port that connects to the U2D2;
```

Shiming-Liang replied:
Ah, sorry I missed this. I was testing on the arm and forgot to change it back.

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/config/mode_configs.yaml:
This looks almost exactly the same as the [default modes file](https://github.com/Interbotix/interbotix_ros_manipulators/blob/humble/interbotix_ros_xsarms/interbotix_xsarm_control/config/modes.yaml). Is there a reason we're changing the port and arm group's profile_acceleration?

Shiming-Liang replied:
We added this file to when we had two arms running together and needed to set a different port for them. I'll remove this file and also the related parameters in the launch script.

lukeschmitt-tr commented on interbotix_ros_xsarms/examples/interbotix_xsarm_gravity_compensation/launch/interbotix_gravity_compensation.launch.py:
```diff
@@ -0,0 +1,127 @@
+    )
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            'mode_configs',
```
This launch arg isn't being passed down to the xsarm_control launch include.

Shiming-Liang replied:
Since the mode_configs is no longer used, this argument is removed in the following commit.

Shiming-Liang committed:
Minor fixes on the gravity compensation demo package

mode_configs.yaml
- removed

launch script
- removed unused mode configs argument
- removed unused get_interbotix_xsarm_models import

readme
- updated the description about the launch script arguments
- updated the dummy namespace to match the launch script logic: we used the robot name as the namespace
